syft/frameworks/torch/tensors/interpreters/gradients_core.py

```diff
@@ -60,10 +49,117 @@ def __init__(self, tensor):
 
     def __call__(self, grad):
         if self.tensor.grad is not None:
-            self.tensor.grad += grad
+            self.tensor.grad += grad.child
         else:
-            self.tensor.grad = grad + 0
+            self.tensor.grad = grad.child.copy()
         return ()
 
     def __repr__(self):
         return self.__class__.__name__
+
+
+def get_mismatch_dims(
+    x_shape, y_shape, x_squash_dims, x_squash_keep_dims, y_squash_keep_dims, current_dim=None
+):
+    """
+    Given two tensors shape, build three lists referencing specific dimension indices where
+    the dimension for the two tensors is not the same:
+
+    Note that we assume len(x_shape) <= len(y_shape) and inverse x and y if necessary.
+
+    1. If one tensor has a longer shape (so lives in higher dimension)(so it is y because len(x_shape)
+    <= len(y_shape)), then all the extra dimensions indices end up in x_squash_dims. Indeed, x will be
+    automatically expanded in all the extra dimensions at forward operation time to match the tensor with
+    the highest dimension. Example: tensor([1]) + tensor([[2], [3]]) is rewritten by torch as
+    tensor([[1], [1]]) + tensor([[2], [3]]). So we need to remove all these dimensions of expansion to
+    get back the gradient of x.
+
+    2. If x or y has a 1 in some dimension and doesn't match the other tensor (ex: torch.Size([3, 7, 5]) and
+    torch.Size([3, 1, 5])), then when an operation is called the tensor is expanded on this dimension, so we
+    need to register this information to do the same squashing as before but we keep the dimension this time.
+    This apply to both tensors and hence we use the lists x_squash_keep_dims and y_squash_keep_dims
+
+    3. current_dim is used to keep count of the dimension, as we iterate on the dimension starting from the
+    highest (the most right one) down to zero. Generally, remember that the shape tuple should match when
+    align on the right, unless there is a 1 instead of the other value: shapes (5,4,3,2) and (4,1,2) are
+    compatible, (5,4,3,2) and (5,4,2,2) are not.
+    """
+    # Initialize the dimension if needed
+    if current_dim is None:
+        current_dim = max(len(x_shape), len(y_shape)) - 1
+        # Inverse x and y if needed to satisfy len(x_shape) <= len(y_shape)
+        if len(x_shape) > len(y_shape):
+            return get_mismatch_dims(
+                y_shape, x_shape, x_squash_dims, x_squash_keep_dims, y_squash_keep_dims, current_dim
```

Comment:
Doesn't it reduce twice the current_dim to call the function again instead of just swapping x_shape and y_shape?

Reply:
No I don't think so because I call it with a return

Reply:
Oh yes, my bad, I thought you were doing something like `current_dim -= 1` just above. It's all good!